bdai_ros2_wrappers/test/test_service_handle.py

```diff
@@ -0,0 +1,250 @@
+#  Copyright (c) 2023 Boston Dynamics AI Institute LLC. All rights reserved.
+import unittest
+from threading import Thread
+from typing import Callable, Optional, Tuple
+
+import rclpy
+from rclpy.callback_groups import ReentrantCallbackGroup
+from rclpy.executors import ExternalShutdownException, MultiThreadedExecutor
+from rclpy.node import Node
+from std_srvs.srv import Empty, SetBool, Trigger
+
+from bdai_ros2_wrappers.service_handle import ServiceHandle
+from bdai_ros2_wrappers.type_hints import Srv, SrvTypeRequest, SrvTypeResponse
+
+
+def _callback_trigger_success(req: SrvTypeRequest, resp: SrvTypeResponse) -> SrvTypeResponse:
+    result = Trigger.Response()
+    result.success = True
+    result.message = "foo"
+    return result
+
+
+def _callback_trigger_fail(req: SrvTypeRequest, resp: SrvTypeResponse) -> SrvTypeResponse:
+    result = Trigger.Response()
+    result.success = False
+    result.message = "bar"
+    return result
+
+
+def _callback_setbool_response(req: SrvTypeRequest, resp: SrvTypeResponse) -> SrvTypeResponse:
+    result = SetBool.Response()
+    if req.data:
+        result.success = True
+        result.message = "foo"
+    else:
+        result.success = False
+        result.message = "bar"
+    return result
+
+
+def _callback_empty_service(req: SrvTypeRequest, resp: SrvTypeResponse) -> SrvTypeResponse:
+    result = Empty.Response()
+    return result
+
+
+class FooBarService(Node):
+    def __init__(self, service_type: Srv, service_name: str, callback: Callable) -> None:
+        super().__init__("foobar_server")
+        self._service = self.create_service(service_type, service_name, callback=callback)
+        self._executor: Optional[MultiThreadedExecutor] = MultiThreadedExecutor(3)
+        self._executor.add_node(self)
+        self._keep_spinning = False
+        self._thread: Optional[Thread] = Thread(target=self._spin, daemon=True)
+        self._thread.start()
+
+    def _spin(self) -> None:
+        if self._executor is not None:
+            self._keep_spinning = True
+            try:
+                while self._context.ok() and self._keep_spinning:
+                    self._executor.spin_once(timeout_sec=0.1)
+            except (ExternalShutdownException, KeyboardInterrupt):
+                pass
+            self._keep_spinning = False
+
+    def stop(self) -> None:
+        self._keep_spinning = False
+        if self._executor is not None:
+            self._executor.shutdown()
+            self._executor.remove_node(self)
+            self._executor = None
+        if self._thread is not None:
+            self._thread.join()
+            self._thread = None
+        self._service = None
+
+
+class ServiceHandleTriggerTest(unittest.TestCase):
```

Comment:
Its nice to have a human readable description of what exactly each test is checking for.

Reply:
Will include